File: go/pkg/minimax/task.go
package minimax

import (
	"context"
	"fmt"
	"net/url"
	"time"
)

// taskType identifies the type of async task for polling.
type taskType int

const (
	taskTypeVideo taskType = iota
	taskTypeSpeechAsync
)

// Task represents an async operation that can be polled for completion.
type Task[T any] struct {
	// ID is the task identifier.
	ID string

	client   *Client
	taskType taskType
}

// NewVideoTask creates a Task for querying an existing video generation task.
func (c *Client) NewVideoTask(taskID string) *Task[VideoResult] {
	return &Task[VideoResult]{
		ID:       taskID,
		client:   c,
		taskType: taskTypeVideo,
	}
}

// NewSpeechAsyncTask creates a Task for querying an existing async speech task.
func (c *Client) NewSpeechAsyncTask(taskID string) *Task[SpeechAsyncResult] {
	return &Task[SpeechAsyncResult]{
		ID:       taskID,
		client:   c,
		taskType: taskTypeSpeechAsync,
	}
}

// Wait waits for the task to complete and returns the result.
//
// Uses a default polling interval of 5 seconds. Use WaitWithInterval
// for custom intervals.
//
// Example:
//
//	ctx, cancel := context.WithTimeout(ctx, 5*time.Minute)
//	defer cancel()
//	result, err := task.Wait(ctx)
func (t *Task[T]) Wait(ctx context.Context) (*T, error) {
	return t.WaitWithInterval(ctx, 5*time.Second)
}

// WaitWithInterval waits for the task to complete with a custom polling interval.
func (t *Task[T]) WaitWithInterval(ctx context.Context, interval time.Duration) (*T, error) {
	// Query immediately before first ticker interval
	result, status, err := t.query(ctx)
	if err != nil {
		return nil, err
	}
	switch status {
	case TaskStatusSuccess:
		return result, nil
	case TaskStatusFailed:
		return nil, fmt.Errorf("task %s failed with status %s", t.ID, status)
	}

	ticker := time.NewTicker(interval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-ticker.C:
			result, status, err := t.query(ctx)
			if err != nil {
				return nil, err
			}

			switch status {
			case TaskStatusSuccess:
				return result, nil
			case TaskStatusFailed:
				return nil, fmt.Errorf("task %s failed with status %s", t.ID, status)
			case TaskStatusPending, TaskStatusQueueing, TaskStatusPreparing, TaskStatusProcessing:
				// Continue waiting
			default:
				return nil, fmt.Errorf("unknown task status: %s", status)
			}
		}
	}
}

// Status returns the current task status without blocking.
func (t *Task[T]) Status(ctx context.Context) (TaskStatus, error) {
	_, status, err := t.query(ctx)
	return status, err
}

// query queries the task status and result.
func (t *Task[T]) query(ctx context.Context) (*T, TaskStatus, error) {
	switch t.taskType {
	case taskTypeVideo:
		return t.queryVideoTask(ctx)
	case taskTypeSpeechAsync:
		return t.querySpeechAsyncTask(ctx)
	default:
		return nil, "", fmt.Errorf("unknown task type")
	}
}

// queryVideoTask queries a video generation task.
func (t *Task[T]) queryVideoTask(ctx context.Context) (*T, TaskStatus, error) {
	var resp struct {
		TaskID      string     `json:"task_id"`
		Status      TaskStatus `json:"status"`
		FileID      string     `json:"file_id,omitempty"`
		VideoWidth  int        `json:"video_width,omitempty"`
		VideoHeight int        `json:"video_height,omitempty"`
		BaseResp    *baseResp  `json:"base_resp,omitempty"`
	}

	// Use query parameter instead of path parameter
	err := t.client.http.request(ctx, "GET", "/v1/query/video_generation?task_id="+url.QueryEscape(t.ID), nil, &resp)
	if err != nil {
		return nil, "", err
	}

	if resp.Status == TaskStatusSuccess {
		// Get download URL from file retrieve API
		downloadURL := ""
		if resp.FileID != "" {
			var fileResp struct {
				File struct {
					DownloadURL string `json:"download_url"`
				} `json:"file"`
				BaseResp *baseResp `json:"base_resp,omitempty"`
			}
			fileErr := t.client.http.request(ctx, "GET", "/v1/files/retrieve?file_id="+url.QueryEscape(resp.FileID), nil, &fileResp)
			if fileErr == nil {
				downloadURL = fileResp.File.DownloadURL
			}
		}

		result := any(&VideoResult{
			FileID:      resp.FileID,
			DownloadURL: downloadURL,
			VideoWidth:  resp.VideoWidth,
			VideoHeight: resp.VideoHeight,
		})
		return result.(*T), resp.Status, nil
	}

	return nil, resp.Status, nil
}

// querySpeechAsyncTask queries an async speech task.
func (t *Task[T]) querySpeechAsyncTask(ctx context.Context) (*T, TaskStatus, error) {
	var resp struct {
		TaskID    int64      `json:"task_id"`
		Status    TaskStatus `json:"status"`
		FileID    int64      `json:"file_id,omitempty"`
		ExtraInfo *AudioInfo `json:"extra_info,omitempty"`
		Subtitle  *Subtitle  `json:"subtitle,omitempty"`
		BaseResp  *baseResp  `json:"base_resp,omitempty"`
	}

	// Use query parameter: /v1/query/t2a_async_query_v2?task_id=xxx
	err := t.client.http.request(ctx, "GET", "/v1/query/t2a_async_query_v2?task_id="+url.QueryEscape(t.ID), nil, &resp)
	if err != nil {
		return nil, "", err
	}

	if resp.Status == TaskStatusSuccess {
		result := any(&SpeechAsyncResult{
			FileID:    fmt.Sprintf("%d", resp.FileID),
			AudioInfo: resp.ExtraInfo,
			Subtitle:  resp.Subtitle,
		})
		return result.(*T), resp.Status, nil
	}

	return nil, resp.Status, nil
}

File: rust/minimax/src/task.rs
//! Async task management.

use std::{sync::Arc, time::Duration};

use serde::{Deserialize, Serialize};

use super::{
    error::{Error, Result},
    http::HttpClient,
    speech::SpeechAsyncResult,
    types::{BaseResp, TaskStatus},
    video::VideoResult,
};

/// Task type.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TaskType {
    SpeechAsync,
    Video,
    VideoAgent,
}

/// An async task that can be polled for completion.
pub struct Task {
    id: String,
    task_type: TaskType,
    http: Arc<HttpClient>,
}

impl Task {
    /// Creates a new task.
    pub fn new(id: String, task_type: TaskType, http: Arc<HttpClient>) -> Self {
        Self { id, task_type, http }
    }

    /// Returns the task ID.
    pub fn id(&self) -> &str {
        &self.id
    }

    /// Returns the task type.
    pub fn task_type(&self) -> TaskType {
        self.task_type
    }

    /// Queries the current status of the task.
    pub async fn status(&self) -> Result<TaskStatusResponse> {
        match self.task_type {
            TaskType::SpeechAsync => self.query_speech_status().await,
            TaskType::Video => self.query_video_status().await,
            TaskType::VideoAgent => self.query_video_agent_status().await,
        }
    }

    /// Waits for the task to complete and returns the result.
    ///
    /// This method polls the task status every 2 seconds until it completes.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// let task = client.speech().create_async_task(&request).await?;
    /// let result = task.wait().await?;
    ///
    /// match result {
    ///     TaskResult::Speech(speech) => {
    ///         println!("File ID: {}", speech.file_id);
    ///     }
    ///     _ => {}
    /// }
    /// ```
    pub async fn wait(&self) -> Result<TaskResult> {
        self.wait_with_interval(Duration::from_secs(2)).await
    }

    /// Waits for the task to complete with a custom polling interval.
    pub async fn wait_with_interval(&self, interval: Duration) -> Result<TaskResult> {
        loop {
            let status = self.status().await?;

            match status.status {
                TaskStatus::Success => {
                    return Ok(status
                        .result
                        .ok_or_else(|| Error::TaskFailed("no result returned".to_string()))?);
                }
                TaskStatus::Failed => {
                    return Err(Error::TaskFailed(
                        status.error_message.unwrap_or_else(|| "unknown error".to_string()),
                    ));
                }
                _ => {
                    tokio::time::sleep(interval).await;
                }
            }
        }
    }

    async fn query_speech_status(&self) -> Result<TaskStatusResponse> {
        #[derive(Serialize)]
        struct Request<'a> {
            task_id: &'a str,
        }

        #[derive(Deserialize)]
        struct Response {
            status: TaskStatus,
            #[serde(default)]
            file_id: String,
            extra_info: Option<super::types::AudioInfo>,
            subtitle: Option<super::types::Subtitle>,
            #[serde(default)]
            base_resp: Option<BaseResp>,
        }

        let resp: Response = self
            .http
            .request(
                "POST",
                "/v1/t2a_async/fetch",
                Some(&Request { task_id: &self.id }),
            )
            .await?;

        let result = if resp.status == TaskStatus::Success {
            Some(TaskResult::Speech(SpeechAsyncResult {
                file_id: resp.file_id,
                audio_info: resp.extra_info,
                subtitle: resp.subtitle,
            }))
        } else {
            None
        };

        Ok(TaskStatusResponse {
            task_id: self.id.clone(),
            status: resp.status,
            result,
            error_message: None,
        })
    }

    async fn query_video_status(&self) -> Result<TaskStatusResponse> {
        let path = format!("/v1/query/video_generation?task_id={}", self.id);

        #[derive(Deserialize)]
        struct Response {
            status: TaskStatus,
            #[serde(default)]
            file_id: String,
            #[serde(default)]
            video_width: i32,
            #[serde(default)]
            video_height: i32,
            #[serde(default)]
            base_resp: Option<BaseResp>,
        }

        let resp: Response = self.http.request::<(), _>("GET", &path, None).await?;

        let result = if resp.status == TaskStatus::Success {
            Some(TaskResult::Video(VideoResult {
                file_id: resp.file_id,
                video_width: resp.video_width,
                video_height: resp.video_height,
                download_url: None,
            }))
        } else {
            None
        };

        Ok(TaskStatusResponse {
            task_id: self.id.clone(),
            status: resp.status,
            result,
            error_message: None,
        })
    }

    async fn query_video_agent_status(&self) -> Result<TaskStatusResponse> {
        let path = format!("/v1/query/video_agent?task_id={}", self.id);

        #[derive(Deserialize)]
        struct Response {
            status: TaskStatus,
            #[serde(default)]
            file_id: String,
            #[serde(default)]
            download_url: String,
            #[serde(default)]
            base_resp: Option<BaseResp>,
        }

        let resp: Response = self.http.request::<(), _>("GET", &path, None).await?;

        let result = if resp.status == TaskStatus::Success {
            Some(TaskResult::Video(VideoResult {
                file_id: resp.file_id,
                video_width: 0,
                video_height: 0,
                download_url: Some(resp.download_url),
            }))
        } else {
            None
        };

        Ok(TaskStatusResponse {
            task_id: self.id.clone(),
            status: resp.status,
            result,
            error_message: None,
        })
    }
}

/// Response from querying task status.
#[derive(Debug, Clone)]
pub struct TaskStatusResponse {
    /// Task ID.
    pub task_id: String,

    /// Current status.
    pub status: TaskStatus,

    /// Result (if completed successfully).
    pub result: Option<TaskResult>,

    /// Error message (if failed).
    pub error_message: Option<String>,
}

/// Result of a completed task.
#[derive(Debug, Clone)]
pub enum TaskResult {
    Speech(SpeechAsyncResult),
    Video(VideoResult),
}
